lkcrypto.c

```diff
@@ -213,6 +262,85 @@ static int decrypt_encrypt(lua_State *L, int dec_enc)
 	}
 }
 
+static int decrypt_encrypt_aead(lua_State *L, int dec_enc)
+{
+	size_t key_len = 0;
+	size_t iv_len = 0;
+	size_t auth_len = 0;
+	size_t ciphertext_len = 0;
+	size_t aad_len = 0;
+
+	int result = 0;
+	int nargs = 0;
+	const u8 *key;
+	const u8 *iv_orig;
+	const u8 *auth_orig;
+	const u8 *aad_orig;
+	const u8 *ciphertext;
+	u8 iv[AES_MAX_KEY_SIZE] = {
+		0,
+	};
+	u8 auth[MAX_AUTH_SIZE] = {
+		0,
+	};
+	u8 *buffer = NULL;
+	u8 *aad = NULL;
+
+	struct cipher_aead_tfm_t *tfm = NULL;
+
+	nargs = lua_gettop(L);
+
+	tfm = luaL_checkudata(L, 1, CUJO_CIPHER_AEAD_TFM_LUA_NAME);
+	key = luaL_checklstring(L, 2, &key_len);
+	ciphertext = luaL_checklstring(L, 3, &ciphertext_len);
+	iv_orig = luaL_checklstring(L, 4, &iv_len);
+	memcpy(iv, iv_orig, iv_len);
+	if (dec_enc == 0) {
+		auth_orig = luaL_checklstring(L, 5, &auth_len);
+		memcpy(auth, auth_orig, auth_len);
+	}
+
+	buffer = kmalloc(ciphertext_len, GFP_ATOMIC);
+
+	if (dec_enc == 1) {
+		auth_len = key_len;
+		if (nargs == 5) {
+			aad_orig = luaL_checklstring(L, 5, &aad_len);
+			aad = kmalloc(aad_len, GFP_ATOMIC);
+			memcpy(aad, aad_orig, aad_len);
+		}
+	}
+
+	memcpy(buffer, ciphertext, ciphertext_len);
+
+	result = enc_dec_aead(tfm, key, key_len, iv, iv_len, auth, auth_len,
+			      aad, aad_len, buffer, ciphertext_len, dec_enc);
+
+	if (!result) {
+		if (dec_enc == 1) {
+			lua_pushlstring(L, buffer, ciphertext_len);
+			lua_pushlstring(L, auth, auth_len);
+			kfree(buffer);
+			if (aad_len)
+				kfree(aad);
+			return 2;
+		} else {
+			lua_pushlstring(L, buffer, ciphertext_len);
+			kfree(buffer);
+			return 1;
+		}
+	} else {
```

Comment:
To avoid a double negative here (`!result` .. `else`), it'd be nicer to check `if (result)` first. Then could also reduce indentation for the success case.